### PR TITLE
Add YAML-metadata for notes and change the document title accordingly

### LINKED ISSUE #303: Change document title in editor
**Which part of the project should be enhanced?**
The editor

**Describe the solution you'd like**
Change the document title to show the title of the note

### PR DESCRIPTION
**Pull request by DerMolly**
### Description
This PR changed the title of the page according to the yaml metadata in the current document.

### Steps

- [x] added implementation
- [x] added / updated tests
- [x] added / updated documentation
- [x] extended changelog

### Related Issue(s)
closes #303

### CODE REVIEW

**Comment by ErikMichelson** [comment]
I suggest the following fallback-strategy (in order) for the title:
1. Use the `title` yaml-metadata
2. Use the `opengraph: title` yaml-metadata
3. Use the first heading of the document
4. something static like "Untitled"

**Comment by DerMolly** [comment]
also relevant https://github.com/codimd/server/issues/69

**Comment by DerMolly** [comment]
I'm currently thinking about searching for the first line that starts with `# ` (notice the space). So the first level one heading. I'm currently debating if I should just operate on the markdown or use the markdown to html parser in the markdown-parser to be more confident to handle all edge cases…
What do you think @codimd/frontend ?

**Comment by mrdrogdrog** [comment]
I would do that in the markdown parsing. Aka: a markdown it core rule. Because then you'll get a html independent token.

**Comment by DerMolly** [comment]
After #298 is merged this should get e2e tests for the title

File: src/components/editor/editor.tsx [hunk 1]
```diff
@@ -62,22 +62,37 @@ https://vimeo.com/23237102
...
 ## Code highlighting
 \`\`\`javascript=
+
 let a = 1
 \`\`\`
 
 `)
   const isWide = useMedia({ minWidth: 576 })
   const [firstDraw, setFirstDraw] = useState(true)
+  const [documentTitle, setDocumentTile] = useState('Untitled')
```
Comment (ErikMichelson) [inline src/components/editor/editor.tsx]:
```suggestion
  const [documentTitle, setDocumentTitle] = useState('Untitled')
```

This needs proper renaming through the code as the set-method is used multiple times.

Comment (DerMolly) [reply src/components/editor/editor.tsx]: ups. fixed that

File: src/components/editor/markdown-renderer/markdown-renderer.tsx [hunk 2]
```diff
@@ -151,17 +219,28 @@ export const MarkdownRenderer: React.FC<MarkdownRendererProps> = ({ content, cla
...
     return ReactHtmlParser(html, { transform: transform })
-  }, [content, markdownIt])
+  }, [content, markdownIt, onMetaDataChange])
 
   return (
     <div className={`markdown-body ${className || ''} d-flex flex-column align-items-center ${wide ? 'wider' : ''}`}>
+      <ShowIf condition={yamlError}>
+        <Alert variant='warning' dir='auto'>
+          <Trans i18nKey='editor.invalidYaml'>
+            <InternalLink text='yaml-metdata' href='/n/yaml-metadata'/>
```
Comment (ErikMichelson) [inline src/components/editor/markdown-renderer/markdown-renderer.tsx:240]:
```suggestion
            <InternalLink text='yaml-metadata' href='/n/yaml-metadata'/>
```

File: src/components/editor/markdown-renderer/markdown-renderer.tsx [hunk 3]
```diff
@@ -68,6 +98,31 @@ export const MarkdownRenderer: React.FC<MarkdownRendererProps> = ({ content, cla
       langPrefix: '',
       typographer: true
     })
+    if (onFirstHeadingChange) {
+      md.core.ruler.after('normalize', 'extract first L1 heading', (state) => {
+        const lines = state.src.split('\n')
+        for (const line of lines) {
+          if (line.startsWith('# ')) {
+            firstHeadingRef.current = line.replace('# ', '').replace(/!?\[([^\]]*)]\([^)]*\)/, '$1')
```
Comment (mrdrogdrog) [inline src/components/editor/markdown-renderer/markdown-renderer.tsx:106]: Extract the regex into a const

Comment (mrdrogdrog) [reply src/components/editor/markdown-renderer/markdown-renderer.tsx]: solved it myself

**Comment by ErikMichelson** [comment]
Maybe we should care about the failing pipeline? It looks as the e2e-test only fails due to a 5-second timeout.

**Comment by mrdrogdrog** [comment]
@ErikMichelson , yes but the problems occurs in multiple PRs. I created an issue https://github.com/codimd/react-client/issues/343

**Pull request merged**